src/day3.cpp

```diff
@@ -0,0 +1,61 @@
+#include <iostream>
+#include <vector>
+#include <fstream>
+
+std::vector<std::string> readPuzzleInputFromFile(std::string fileName) {
+    std::ifstream inFile(fileName);
+    std::string line;
+    std::vector<std::string> linesInFile{};
+
+    while (getline(inFile, line)) {
+        linesInFile.push_back(line);
+    }
+    return linesInFile;
+}
+
+int numberOfTreesOnSlope(const std::vector<std::string>& area, std::vector<int> slope){
+    int index = 0;
+    int counter = 0;
+    int stepRight = slope[0];
+    int stepDown = slope[1];
+    for (std::vector<std::string>::const_iterator it = area.begin(); it != area.end() + stepDown - 1; it += stepDown) {
+            if ((*it)[index % (*it).length()] == '#') {
+                counter++;
+            }
+            index += stepRight;
+        }
+    return counter;
+}
+
+int productOfTreesOnSlopes(const std::vector<std::string>& inputLines, const std::vector<std::vector<int>>& slopes) {
+    
+    std::vector<int> numberOfTreesPerSlope{};
+
+    for (std::vector<std::vector<int>>::const_iterator it = slopes.begin(); it != slopes.end(); it++) {
+        int numberOfTrees = numberOfTreesOnSlope(inputLines, *it);
+        numberOfTreesPerSlope.push_back(numberOfTrees);
```

Comment:
Using `.push_back` on a vector is a bit of shame if you know the size beforehand. What happens to a vector under the hood is that it doesn't know how big it's going to be, so it will choose *some* size, then if you call `.push_back` while its memory is full (the maximum memory is called 'capacity'), it will copy all the data to a place in memory that is twice as big. This is brilliant, because the time complexity of this is still O(n) for n insertions because the capacity is constantly doubled (its need to do work - the copy - but the amount of time this is needed becomes less and less, so in the limit the complexity is O(n) - this is called amortized cost). But... still it needs to do some copies! But we already know how big it is going to be before entering the loop: namely `slopes.size()`.

So what can be do? Two things: set the capacity of `numberOfTreesPerSlope` before the loop, like this:

```
numberOfTreesPerSlope.reserve(slopes.size());

// Note: it now has the capacity (so the memory) for `slopes.size()` entries, but note that the *size* is still 0!
```

OR you can directly set the size, using `.resize` or directly in the constructor:

```
std::vector<int> numberOfTreesPerSlope(slopes.size());

// Now the *size* is actually `numberOfTreesPerSlope`! Note however that the data inside is uninitialized, so don't use it yet because you'll get undefined behavior!
```

But now you can no longer use `push_back`, but instead need to index the items, e.g. `numberOfTreesPerSlope[i] = `.

I kind of prefer the latter, because it's more explicit, but you'll need an index to iterate over `slopes` and `numberOfTreesPerSlope` at the same time, i.e.:

```
for(std::size_t i = 0; i < slopes.size(); ++i)
{
     int numberOfTrees = numberOfTreesOnSlope(inputLines, slopes[i]);
     numberOfTreesPerSlope[i] = numberOfTrees;
}
```